esmvalcore/_recipe.py

```diff
@@ -1552,12 +1622,26 @@ def _fill_wildcards(self, variable_group, preprocessor_output):
                 variable_group].update(
                     {'additional_datasets': updated_datasets})
 
-    def _create_preprocessor_tasks(self, diagnostic_name, diagnostic):
+    def _create_preprocessor_tasks(self, diagnostic_name, diagnostic,
+                                   tasknames_to_run, any_diag_script_is_run):
         """Create preprocessor tasks."""
         tasks = []
         failed_tasks = []
         for variable_group in diagnostic['preprocessor_output']:
             task_name = diagnostic_name + TASKSEP + variable_group
+
+            # Skip preprocessor if not a single diagnostic script is run and
+            # the preprocessing task is not explicitly requested by the user
+            if tasknames_to_run:
```

Comment:
why wouldn't `tasknames_to_run` be empty in this case? ie no `preprocessor` and no `diagnostics`

Reply:
Empty `tasknames_to_run` is the default value used when `--diagnostics` is not used; in this case, all tasks are run. But I agree, that feels weird, but much simpler to code.

Reply:
okies :+1: